torchtune/data/_templates.py

```diff
@@ -5,7 +5,9 @@
 # LICENSE file in the root directory of this source tree.
 
 from abc import ABC, abstractmethod
-from typing import Any, Dict, Mapping, Optional
+from typing import Dict, Optional
+
+from torchtune.data._types import Sample
```

Comment:
Nit: What does creating a new datatype add for us? 

`Mapping[str, Any]` is common and makes sense, whereas the notion of a sample is something new entirely. If the value add is really tremendous, this makes sense, but in the spirit of our design principles to keep things as simple as possible, I want us to be conscious about this decision.

Reply:
I can concede on Sample since it's more of a cosmetic change, but the other types are needed to enforce a standardized conversation format and roles.